### ISSUE TITLE
`--efficient-long-context` silently corrupts SFT loss above seq 512 and goes NaN at seq >= 2048 (Qwen3.5-4B bf16, `--train-type full`)

### ISSUE THREAD

**Issue by writerblockade** [Goekdeniz-Guelmez/mlx-lm-lora#84]
Hi — thanks for mlx-lm-lora. It is the reason full-parameter finetuning of a hybrid-linear model is possible on this machine at all, and it is what the mlx-lm #1206 thread points people at. This report comes out of a training-path calibration ladder for a Danish finetune of Qwen3.5-4B-Base: four trainers × five sequence lengths, every cell logged, run on one machine over one night. `--efficient-long-context` was the only thing that produced wrong numbers, and it produced them quietly, so it seemed worth writing up properly rather than just working around.

### Summary

With `--train-type full` on Qwen3.5-4B-Base (bf16), `--efficient-long-context` degrades the training loss as sequence length grows, and the degradation is **silent** — the run looks healthy, the progress bar advances, memory and throughput are reported normally:

| max_seq_length | with `--efficient-long-context` | same command, flag removed | reference (stock mlx-lm, identical data/seed/LR) |
|---:|---|---|---|
| 512 | loss 4.054 → 1.844, looks correct | — | 4.158 → 1.634 |
| 1024 | **starts at 6.500**, sits at 5.203 | — | **starts at 4.264**, 2.416 at iter 90 |
| 2048 | **NaN from the first report**, and never recovers | **3.793 → 2.236, clean** | 3.238 at iter 20 (112GB, near OOM) |
| 4096 | **NaN from the first report** | OOM at 16s | OOM |

Two things I would flag as the actual bug, separately:

1. **At seq 1024 there is no NaN and no error — the loss is just wrong.** It starts ~2.2 nats above where the identical data/seed/LR starts under stock mlx-lm and stays there. A user running only this configuration would see a plausible-looking curve descending from 6.5 and never know. That is the part that cost me the most time.
2. **At seq ≥ 2048 it is NaN from the very first report** (`Iter 10: loss nan`), not a divergence that develops.

And a smaller, separable observation: at seq 2048 the flag is also **slower and heavier** than not using it — 46.0 tok/s in 44.9GB with the flag versus **134.6 tok/s in 41.8GB** without it. So on this model at this length the flag costs throughput and memory *and* correctness. I don't know whether that holds generally, but it made "just drop the flag" an easy call here.

### Minimal reproduction

Two commands, identical except for the one flag:

```sh
# NaN
mlx_lm_lora.train --model <Qwen3.5-4B-Base, MLX bf16> --train --data <data> \
  --train-type full --train-mode sft --num-layers -1 \
  --batch-size 1 --iters 500 --max-seq-length 2048 \
  --grad-checkpoint --efficient-long-context \
  --learning-rate 1e-5 --seed 1234 \
  --steps-per-report 10 --val-batches 0 --save-every 100000 \
  --adapter-path <scratch>

# clean — same everything, --efficient-long-context removed
mlx_lm_lora.train --model <Qwen3.5-4B-Base, MLX bf16> --train --data <data> \
  --train-type full --train-mode sft --num-layers -1 \
  --batch-size 1 --iters 500 --max-seq-length 2048 \
  --grad-checkpoint \
  --learning-rate 1e-5 --seed 1234 \
  --steps-per-report 10 --val-batches 0 --save-every 100000 \
  --adapter-path <scratch>
```

Observed, first report line of each:

```
# with the flag
Iter 10: loss nan,   lr 1.000e-05, it/s 0.045, tok/s 46.512,  trained_tok 10260.0, peak_mem 44.906GB
Iter 20: loss nan,   lr 1.000e-05, it/s 0.045, tok/s 46.206,  trained_tok 20520.0, peak_mem 44.906GB

# without the flag
Iter 10: loss 3.793, lr 1.000e-05, it/s 0.042, tok/s 85.400,  trained_tok 20470,   peak_mem 41.808GB
Iter 20: loss 3.566, lr 1.000e-05, it/s 0.055, tok/s 113.136, trained_tok 40940,   peak_mem 41.808GB
```

Note `trained_tok` at iter 10: **10260 with the flag vs 20470 without**, for the same nominal 10 iterations at the same `--max-seq-length`. Roughly half the tokens are being accounted for under the flag. I can't tell from outside whether that is a bookkeeping difference or a sign that the chunking is dropping or mis-splitting part of each sequence, but it is the clearest quantitative hint I have, and it is consistent with the loss being wrong at 1024 before it is NaN at 2048.

Also, flat peak memory across lengths under the flag — 44.588GB at 1024, 44.906GB at 2048, 45.040GB at 4096 — is suspicious in the same direction: memory barely moves while the sequence quadruples.

The data was plain SFT-formatted Danish text (public-domain parliamentary records), nothing unusual: 256 train documents, one document per example, no short/degenerate rows. The `Val loss nan` line at iter 1 in the logs is just `--val-batches 0` and is present in the clean runs too — the failure being reported is the **train** loss.

### Environment

| | |
|---|---|
| mlx-lm-lora | 3.1.2 |
| mlx / mlx-metal | 0.32.2 |
| mlx-lm | 0.31.3 |
| transformers | 5.15.1 |
| Python | 3.12.13 |
| machine | Apple M5 Max, 128GB unified memory, macOS 27.0 |
| `iogpu.wired_limit_mb` | 0 (macOS default) at the time of these runs |
| model | `Qwen/Qwen3.5-4B-Base` converted with `mlx_lm convert --dtype bfloat16` |

### What I can supply

Full raw logs for every cell above (with the exact command line in the header of each), the paired flag-on/flag-off runs at seq 2048, and the same ladder for stock `mlx-lm` and PyTorch MPS as controls, if any of that is useful. Happy to run a patch against this setup — reproducing it takes about 15 minutes per cell.

Thanks again for the project.

**Comment by Goekdeniz-Guelmez** [comment]
Hey @writerblockade 
thanks for flagging this issue.
has been fixed in #85

**Comment by writerblockade** [comment]
Thanks for the fast turnaround on #85 — the chunk-boundary reasoning in `iter_cached_sft_chunks` matches what we saw from the outside.

I retested at the merge commit and I'm afraid **the symptom is still present for our configuration**, so you may want to reopen. Same machine, same data, same seed/LR, same command as the original report — only the tool version changed.

**Build:** `mlx-lm-lora` @ `645fb6e36d11031aaa88c2bcf004af790b711031` (the merge of #85 into main), `mlx==0.32.2`, `transformers==5.15.1`, Python 3.12.13, M5 Max 128GB, macOS 27.0. Note the installed package still self-reports `3.1.2`, so the sha is the only identifier.

**Command** (identical to the original report but for the venv):
```
mlx_lm_lora.train --model <Qwen3.5-4B-Base, mlx bf16> --train --data <sft jsonl> \
  --train-type full --train-mode sft --num-layers -1 \
  --batch-size 1 --max-seq-length 1024 --grad-checkpoint --efficient-long-context \
  --learning-rate 1e-5 --seed 1234 --steps-per-report 10 --val-batches 0
```

**seq 1024, `--efficient-long-context` ON:**

| build | iter 10 | iter 20 | trained_tok @10 | peak_mem |
|---|---:|---:|---:|---:|
| v3.1.2, flag ON (before #85) | loss 6.500 | 6.169 | 5130 | 44.588 GB |
| **main @ 645fb6e3, flag ON (after #85)** | **loss 5.863** | **6.059** | **5130** | **90.043 GB** |
| **main @ 645fb6e3, flag OFF** — same build, same data | **loss 4.248** | **3.716** | **10230** | **33.267 GB** |
| stock `mlx-lm` 0.31.3, same data/seed/LR | loss 4.264 | 3.318 | 10230 | 46.851 GB |

Three things, and the flag-off row on the same build is the control for all of them:

1. **`trained_tok` is unchanged at 5130, against 10230 for the same build with the flag off** — which is also exactly what stock `mlx-lm` reports, so this is not a difference in how the two tools count. The chunked path scores **513 targets per step where it should score 1023**: `seq_step_size/2 + 1` rather than `seq_length - 1`. The same shape holds at 2048 — **1025 per step against 2047** — so it is not a constant offset but roughly half at every chunked length. (At `--max-seq-length 512` the flag is a no-op — `batch.shape[1] > seq_step_size` is false — and both report 5110, i.e. `seq_len - 1` per step.)
2. **The loss is rising** (5.863 → 6.059) where the same build with the flag off falls (4.248 → 3.716) and stock falls (4.264 → 3.318). Before #85 it at least fell (6.500 → 6.169), so the trajectory changed — but away from the reference, not toward it.
3. **Peak memory roughly doubled** against the pre-fix flag (44.6 → 90.0 GB) and is **2.7× the same build with the flag off** (33.3 GB) — the opposite of what an "efficient long context" flag is for.

At `--max-seq-length 2048` the patched flag is still **NaN from the first report** (`loss nan` at iters 10 and 20, `trained_tok` 10250/20500, peak 91.1 GB), which is the original #84 symptom unchanged.

At `--max-seq-length 512` the flag is a no-op on both builds (`batch[0].shape[1] > seq_step_size` is false at the default `seq_step_size=512`), and both report `loss 4.063 / 4.060` and `peak_mem 32.751GB` — identical, as expected. So the divergence begins exactly where chunking begins.

**What I checked and could NOT reproduce in isolation.** I wrote a harness that drives `default_loss` through your `iter_cached_sft_chunks` bounds with a live `make_prompt_cache`, and asserts that the chunked masked-token total equals the unchunked one. It **fails on v3.1.2 and passes at 645fb6e3**, on both a plain Llama stack and a synthetic Qwen3.5-style hybrid stack (`full_attention_interval=4`, Gated DeltaNet layers). So **your fix is correct in isolation and my harness can detect the old bug** — whatever is happening in the full training run is not the chunk-bounds arithmetic, and I do not currently have a mechanism to offer you.

I would rather hand you a symptom I can defend than a diagnosis I can't. Happy to run any diagnostic you'd like, and happy to be told this is specific to `--train-type full` on a Gated DeltaNet model rather than general.
